lib/net/ssh/authentication/methods/publickey.rb

```diff
@@ -89,6 +91,22 @@ def authenticate_with(identity, next_service, username)
               raise Net::SSH::Exception, "unexpected reply to USERAUTH_REQUEST: #{message.type} (#{message.inspect})"
             end
           end
+
+          def authenticate_with(identity, next_service, username)
+            alg = identity.ssh_type
+            salg = nil
+            if authenticate_with_2(identity, next_service, username, alg, salg)
+              return true
+            end
+            if !rsa_sha2_auth_disable && alg == "ssh-rsa"
+              # if ssh-rsa fails, retry with sha2
```

Comment:
The RSA with SHA2 should be used by default, not only if the SHA1 fails. This is just a very small subset of RFC 8332. Hardcoding the SHA256 regardless what is negotiated can also cause interoperability issues. 

https://datatracker.ietf.org/doc/html/rfc8332

Reply:
I think certificates should be just ignored currently (SHA2 is not tried for them) as check for RSA is done by comparing ssh_type to "ssh-rsa" and this should fail for certs:
https://github.com/net-ssh/net-ssh/blob/c6a21e5f0a9f0bc92e9c4eb0e17af40d175a97a5/lib/net/ssh/authentication/certificate.rb#L53

Reply:
Do I understand it correctly that userauth request message is the same, just certificate is used instead of key?

Reply:
In that case, using SHA256 is probably fine. I probably forgot about that part of RFC :)

The certificate authentication is using the same messages with different key types. So right, for now, the will be ignored, but you might also add something like this to support them, I think:

```
            if !rsa_sha2_auth_disable && alg == "ssh-rsa-cert-v01@openssh.com"
              # if ssh-rsa certificate fails, retry with sha2
              alg = salg = "rsa-sha2-256-cert-v01@openssh.com"
              return authenticate_with_2(identity, next_service, username,
                alg, salg)
            end
```